File: ios/MapboxNavigationFreeDriveManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(MapboxNavigationFreeDriveManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(onLocationChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onTrackingStateChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onRouteChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(showSpeedLimit, BOOL)
RCT_EXPORT_VIEW_PROPERTY(speedLimitAnchor, NSArray)
RCT_EXPORT_VIEW_PROPERTY(followZoomLevel, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(userPuckImage, UIImage)
RCT_EXPORT_VIEW_PROPERTY(userPuckScale, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(destinationImage, UIImage)
RCT_EXPORT_VIEW_PROPERTY(mapPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(logoVisible, NSArray)
RCT_EXPORT_VIEW_PROPERTY(logoPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(attributionVisible, NSArray)
RCT_EXPORT_VIEW_PROPERTY(attributionPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(lineColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(altLineColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(waypointColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(waypointRadius, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointOpacity, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeWidth, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeOpacity, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(unknownLineColor, NSString)

RCT_EXTERN_METHOD(
  showRouteViaManager: (nonnull NSNumber *)node
  origin: (NSArray *)origin 
  destination: (NSArray *)destination 
  waypoints: (NSArray *)waypoints
  padding: (NSArray *)padding
  styles: (NSDictionaryArray *)styles
  legIndex: (nonnull NSNumber *)legIndex
)

RCT_EXTERN_METHOD(
  clearRouteViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  followViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  moveToOverviewViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  fitCameraViaManager: (nonnull NSNumber *)node
  padding: (NSArray *)padding
)

@end

File: ios/MapboxNavigationFreeDriveManager.swift
@objc(MapboxNavigationFreeDriveManager)
class MapboxNavigationFreeDriveManager: RCTViewManager {
  override func view() -> UIView! {
    return MapboxNavigationFreeDriveView();
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc func showRouteViaManager(_ node: NSNumber, origin: [NSNumber], destination: [NSNumber], waypoints: [[NSNumber]], styles: [NSDictionary], legIndex: NSNumber, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) -> Void {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.showRoute(origin: origin, destination: destination, waypoints: waypoints, styles: styles, legIndex: legIndex, resolver: resolver, rejecter: rejecter)
    }
  }

  @objc func clearRouteViaManager(_ node: NSNumber) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.clearRoute()
    }
  }

  @objc func followViaManager(_ node: NSNumber) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.follow()
    }
  }

  @objc func moveToOverviewViaManager(_ node: NSNumber, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.moveToOverview(padding: padding)
    }
  }

  @objc func fitCameraViaManager(_ node: NSNumber, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.fitCamera(padding: padding)
    }
  }
}
